File: archiver.py
import time
import os
from flask import Flask, jsonify, request
import ipfshttpclient
from ipfshttpclient.exceptions import Error as IPFSError
from git import Repo
from git.exc import GitCommandError

app = Flask(__name__)
repo = Repo('data')

try:
    repo.init()
except GitCommandError as error:
    print(f"git error {str(error)}")

def getIpfs():
    connect = os.environ.get('IPFS_CONNECT')

    if connect:
        return ipfshttpclient.connect(connect, timeout=20)
    else:
        return ipfshttpclient.connect(timeout=20)

def checkIpfs():
    for i in range(10):
        try:
            ipfs = getIpfs()
            #print(ipfs.id())
            return True
        except:
            print(i, "attempting to connect to IPFS...")
            time.sleep(1)
    return False

@app.route('/api/v1/pin/', methods=['POST'])
def pin():
    try:
        data = request.get_json()

        if not data or 'path' not in data:
            print("Failed to pin data: No path provided")
            return jsonify({'error': 'No path provided'}), 400

        if checkIpfs():
            ipfs = getIpfs()
            res = ipfs.add(data['path'], recursive=True, pin=True, pattern="**")
            cid = res[-1]['Hash']
        else:
            print("IPFS not available")
            return jsonify({'error': 'IPFS not available', 'cid': 'TBD'}), 500
    except IPFSError as error:
        print(f"Failed to pin data {data['path']}: {str(error)}")
        return jsonify({'error': f"Failed to pin data: {str(error)}", 'cid': 'TBD'}), 500

    print(f"pinned {data['path']} to {cid}")
    return jsonify({'cid': cid})

@app.route('/api/v1/commit', methods=['POST'])
def commit():
    data = request.get_json()

    if not data or 'message' not in data:
        return jsonify({'error': 'No message provided'}), 400

    message = data['message']

    try:
        repo.git.add('--all')
        repo.git.commit('-m', message)
        githash = repo.git.rev_parse('HEAD')
    except GitCommandError as error:
        return jsonify({'error': f'Failed to commit changes: {str(error)}'}), 500

    # Replace this with your actual implementation
    return jsonify({'ok':1, 'githash':githash})

@app.route('/api/v1/peg', methods=['GET', 'POST'])
def peg():
    # Replace this with your actual implementation
    return jsonify({'message': 'You reached the /api/v1/peg endpoint'})

if __name__ == '__main__':
    port = int(os.getenv('ARC_PORT', 5115))
    app.run(debug=True, host='0.0.0.0', port=port)

File: archiver.js
const path = require('path');
const axios = require('axios');
const config = require('./config');

async function register(xid, cid) {
    try {
        const response = await axios.post(`${config.archiver}/api/v1/register`, { xid: xid, cid: cid });
        return response.data.txid;
    }
    catch (error) {
        console.error(`register error: ${error}`);
    }
}

async function notarize(xid, cid) {
    try {
        const response = await axios.post(`${config.archiver}/api/v1/notarize`, { xid: xid, cid: cid });
        return response.data.txid;
    }
    catch (error) {
        console.error(`notarize error: ${error}`);
    }
}

async function certify(txid) {
    try {
        const response = await axios.post(`${config.archiver}/api/v1/certify`, { txid: txid });
        return response.data;
    }
    catch (error) {
        console.error(`certify error: ${error}`);
    }
}

async function walletinfo() {
    try {
        const response = await axios.get(`${config.archiver}/api/v1/walletinfo`);
        return response.data;
    }
    catch (error) {
        console.error(`walletinfo error: ${error}`);
    }
}

async function ready() {
    try {
        const response = await axios.get(`${config.archiver}/api/v1/ready`);
        return response.data.ready;
    }
    catch (error) {
        console.error(`ready error: ${error}`);
    }
}

// Function to add all changes, commit, and push
async function commitChanges(event) {
    try {
        const commitMessage = JSON.stringify(event);
        const response = await axios.post(`${config.archiver}/api/v1/commit`, { message: commitMessage });
        const commit = response.data;

        if (commit.error) {
            console.error(`commitChanges error: ${commit.error}`);
        }
        else if (commit.githash) {
            const hash = commit.githash.substring(0, 8);
            console.log(`Commit: ${commitMessage} (${hash})`);
            return commit.githash;
        }
    }
    catch (error) {
        console.error(`commitChanges error: ${error.response.data?.error || error}`);
    }
}

async function pushChanges() {
    try {
        const response = await axios.get(`${config.archiver}/api/v1/push`);
        const push = response.data;

        if (push.error) {
            console.error(`pushChanges error: ${push.error}`);
        }
    }
    catch (error) {
        console.error(`pushChanges error: ${error}`);
    }
}

async function getLogs() {
    try {
        const response = await axios.get(`${config.archiver}/api/v1/logs`);
        const data = response.data;

        if (data.error) {
            console.error(`getLogs error: ${data.error}`);
        }
        else {
            return data.logs;
        }
    }
    catch (error) {
        console.error(`getLogs error: ${error}`);
    }
}

async function pinAsset(xid) {
    try {
        const assetPath = path.join(config.assets, xid);
        const response = await axios.get(`${config.archiver}/api/v1/pin/${assetPath}`);
        const ipfs = response.data;
        return ipfs;
    }
    catch (error) {
        console.error(`pinAsset error: ${error}`);
    }
}

module.exports = {
    certify,
    commitChanges,
    getLogs,
    notarize,
    pinAsset,
    pushChanges,
    ready,
    register,
    walletinfo,
};
